File: ruby/medium/longest_increasing_subsequence.rb
# @param {Integer[]} nums
# @return {Integer}
def length_of_lis(nums)
    return 0 unless nums.length > 0
    memo = Array.new(nums.length).fill(1)
    max = 1
    for i in 0..nums.length-1
        for j in i+1..nums.length-1
            if(nums[j] > nums[i])
                memo[j] = [memo[j], memo[i]+1].max
                max = [max, memo[j]].max
            end
        end
    end
    return max
end

File: javascript/medium/LongestIncreasingSubsequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var lengthOfLIS = function(nums) {
    if(nums === null || nums.length === 0) return 0;
    const memo = new Array(nums.length).fill(1);
    let max = 1;
    for(let i = 0; i < nums.length; i++) {
        for(let j = i+1; j < nums.length; j++) {
            if(nums[j] > nums[i]) {
                memo[j] = Math.max(memo[j], memo[i]+1);
                max = Math.max(memo[j], max);
            }
        }
    }
    return max;
};
